timemachine/fe/mcgregor.py

```diff
@@ -135,45 +126,105 @@ def __init__(self, n_vertices, edges):
-        source is a mapped node. i.e. node_states[source] == NODE_STATE_VISITED_MAPPED. If this function
-        returns True, then the resulting graph is definitely disconnected. If this function
-        returns False, then the resulting graph can still be connected.
-
-        For example, let M be mapped node, D be a demapped node, and U be an unvisited node:
+    def mapping_is_disconnected(
+        self,
+        mapped_nodes: Set[int],
+        unvisited_nodes: Set[int],
+        max_connected_components: Optional[int],
+        min_connected_component_size: int,
+    ):
+        r"""Verify whether mapped nodes can still be connected at least the specified minimum number of connected
+        components. It is assumed that source is a mapped node. i.e. node_states[source] == NODE_STATE_VISITED_MAPPED.
+        If this function returns True, then the resulting graph is definitely disconnected. If this function returns
+        False, then the resulting graph can still be connected.
+
+        For example, let M be mapped node, D be a demapped node, and U be an unvisited node. If
+        `max_connected_components=1`:
 
         1. M-U-M # returns False
         2. M-D-M # returns True
         3. M-M-U # returns False
         4. M-M-D # returns False
 
-        The implementation is adapted from the _plain_bfs() method in:
-        https://networkx.org/documentation/stable/_modules/networkx/algorithms/components/connected.html
-
+        This calls an optimized implementation in mapping_is_disconnected_fast; see mapping_is_disconnected_ref for a
+        simpler-to-understand reference version.
         """
-        unseen = [True] * self.n_vertices
-        unseen[source] = False
-        nextlevel = [source]
-        mapped_count = 1
-
-        if mapped_count == n_mapped_nodes:
-            return False
-
-        while nextlevel:
-            thislevel = nextlevel
-            nextlevel = []
-            for v in thislevel:
-                for w in self.get_neighbors(v):
-                    if unseen[w] and node_states[w] != NODE_STATE_VISITED_DEMAPPED:
-                        if node_states[w] == NODE_STATE_VISITED_MAPPED:
-                            mapped_count += 1
-                            if mapped_count == n_mapped_nodes:
-                                return False
-                        unseen[w] = False
-                        nextlevel.append(w)
-        return True
+
+        # Uncomment to assert result of fast implementation matches the reference
+        # args = (mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size)
+        # assert self.mapping_is_disconnected_fast(*args) == self.mapping_is_disconnected_ref(*args)
+
+        return self.mapping_is_disconnected_fast(
+            mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size
+        )
+
+    def mapping_is_disconnected_fast(
```

Comment:
how much faster is `_fast` relative to `_ref`?

Reply:
can we add some tests just to compare the correctness of the `_fast` version to `_ref` given the increased complexity in logic? - eg. on random graphs, molecules from hif2a set, etc, with arbitrary choices of mapped_nodes, unvisited nodes, max_connected_components, etc.

Reply:
> how much faster is _fast relative to _ref?

The difference is substantial -- from ~20-30% slower than the existing implementation to about 10% faster. (With default settings, running a selection of 200 mappings from public datasets.)

> can we add some tests just to compare the correctness of the _fast version to _ref given the increased complexity in logic?

Will do. My approach thus far had been to uncomment the lines in `mapping_is_disconnected` that compares the fast version with the reference for all pairs that we test, but agreed that we should have an automated test.

Reply:
> The difference is substantial -- from ~20-30% slower than the existing implementation to about 10% faster. (With default settings, running a selection of 200 mappings from public datasets.)

Thanks! FWIW when I looked at optimizing the code, roughly the causes for slowness in the networkx code was:

1. Instantiating a nx.Graph from the list of edges every time (But this can be cached).
2. Inducing a subgraph view is actually pretty fast
3. However, checking for existence of neighbors/edges in the view was very very slow. 

> Will do. My approach thus far had been to uncomment the lines in mapping_is_disconnected that compares the fast version with the reference for all pairs that we test, but agreed that we should have an automated test.

Yeah but given how critical this logic is I think we need to make this bullet proof.

Reply:
Added a property test comparing the fast and reference implementations on random inputs in https://github.com/proteneer/timemachine/pull/1349/commits/c208820d09341e622fb0e5cbb249aa093e9e50a6. (This also led me to discover a bug in the reference version around networkx graph creation when there are single-node components.)